Mkps/diode test template

EtherealMC-Bit opened the pull request:
The idea here is to port over a diode test template from Warp so we can use it in WarpX. Along with this a test needs to be adapted to use the ported diode template.

peterscherpelz committed:
Initial, totally-not-complete, diode setup script

Commit:
Renamed Ar to inert_gas, raised not implemented errors, removed Cs-ionization functionality

Commit:
set up grids in init_base()

Commit:
Made not implemented errors more helpful

Commit:
Used PICMI electrostatic solver in init_solver()

peterscherpelz commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -317,35 +262,72 @@ def init_base(self):
+                            lower_boundary_conditions_particles=['periodic', 'absorbing'],
+                            upper_boundary_conditions_particles=['periodic', 'absorbing'],
+                            moving_window_velocity=None,
+                            warpx_max_grid_size=nx//4
```
For now probably just make `MAX_GRID_SIZE` a new class variable and assign it here? or call it `MAX_GRID_SIZE_FACTOR` and divide nz by that (nz, not nx, I'm guessint is what you want)

peterscherpelz commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -317,35 +262,72 @@ def init_base(self):
+                            bc_zmax='dirichlet',
+                            warpx_potential_hi_x=self.V_ANODE,
+                            warpx_potential_lo_x=self.V_CATHODE,
+                            lower_boundary_conditions_particles=['absorbing', 'periodic', 'absorbing'],
```
periodic - periodic - absorbing

peterscherpelz reviewed (commented):
Definitely the right track @MKieburtz

Commit:
Boundry condition parameter fix and MAX_GRID_SIZE_FACTOR field

Commit:
Turned two fields into class input variables and added init_warpx() function

- Turned diag_steps and total_timesteps into class input variables called
  DIAG_STEPS and TOTAL_TIMESTEPS.

Commit:
commented out unused gentools.SetupCalcs() function

Commit:
added sim_control variable

peterscherpelz commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -343,43 +348,38 @@ def init_base(self):
-            self.solver.mgmaxiters = 10000
-
-        warp.registersolver(self.solver)
+        elif self.dim == 2 or self.dim == 3:
```
Python gives you shorthand `elif self.dim in [2, 3]:`

peterscherpelz reviewed (commented):
@MKieburtz This is looking good! Next steps could include:
- Eliminate old `metools` imports.
- Move `mewarpx/tests/test_runsuite1d.py` to `mewarpx/tests/test_capacitive_discharge.py` and adapt to running the capacitive discharge simulation. Running it in 2D, 3D, and RZ would be great.
- See how `parametrize` is currently used for 1D vs 2D in that testfile; reuse that for 2D vs 3D vs RZ.
- You'll probably need to do the _libwarpx import inside the test function in that testfile once you know the dimensionality.
- You'll probably want to copy `initialize_testingdir` over from warp for the test file. See https://github.com/ModernElectron/ME_warp/blob/master/metools/metools/init_restart_util.py#L209

Commit:
removed old imports

Commit:
Attempt at running capacitive_discharge using diode_setup

peterscherpelz commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -284,24 +277,26 @@ def init_base(self):
 
-        nz = int(round((zmmax - zmmin)/self.RES_LENGTH))
+        if self.NZ is None:
+            self.NZ = int(round((zmmax - zmmin)/self.RES_LENGTH))
```
I think it would be good to check and force that if NX, NY, and NZ are all not None as inputs, RES_LENGTH must be None. Then compute RES_LENGTH based on (zmmax-zmmin)/NZ.

peterscherpelz commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -626,10 +621,25 @@ def init_resultsinfo(self):
-    def init_warpx(self):
-        sim = picmi.Simulation(
+    def init_simulation(self, **kw):
+        warpx_collisions = kw.pop('warpx_collisions')
```
`warpx_collisions` should be what's created in the `inert_gas_ionization` function above. So that should be integrated into this rather than being a passed kw. If that function isn't run, defaulting warpx_collisions to None should be fine.

peterscherpelz commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -626,10 +621,25 @@ def init_resultsinfo(self):
\ No newline at end of file
+        mwxrun.init_run(simulation=self.sim)
+
+    def add_species_to_sim(self, species, particle_per_cell):
```
This should be able to happen with input parameters like most things in here too, rather than called separately. I think in Warp it was part of the setup injectors area, but we can move stuff around.

peterscherpelz commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,366 @@
+
+constants = picmi.constants
+
+##########################
```
All parameters should be inside the test function - only imports and global variables across all tests belong outside. Hopefully most of these can be directly passed to the DiodeRun setup rather than being declared separately.

EtherealMC-Bit replied:
Some of the parameters make it clearer how the values of certain things like DT are obtained. For example `DT = 1.0 / (400 * FREQ)` is better than `DT = 1.0 / (400 * 13.56e6)`. `N_INERT` and `T_INERT` are used to create the uniform distributions so they seem to make sense as variables as well to me.

peterscherpelz commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,366 @@
+    basename = "Run1D_alldiags"
+    # Include a random run number to allow parallel runs to not collide. Using
+    # python randint prevents collisions due to numpy rseed below
+    #initialize_testingdir(name)
```
We do want to keep this function.

peterscherpelz commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,366 @@
+@pytest.mark.parametrize(
+    ("name"),
+    [
+        'Run1D_alldiags_2DScraper',
```
Change these names - alldiags is no longer relevant, but make these descriptive 2D vs RZ vs 3D

peterscherpelz commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,366 @@
+        init_resultsinfo=True
+    )
+
+    sim = run.init_simulation(warpx_collisions=[mcc_electrons, mcc_ions])
```
So MCC, and adding species to sim, should all be integrated into the setup.

peterscherpelz commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,366 @@
+    run.add_species_to_sim(ions, particle_per_cell=number_per_cell_each_dim)
+
+    run.init_warpx(sim=sim)
+    # if use_2d_scraper:
```
Cut out PHistDiag section entirely

peterscherpelz commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,366 @@
+
+    # warputil.warp_generate()
+
+    # # Run the main WARP loop
```
We do want to run the simulation in this manner with Nick's control loop.

peterscherpelz commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,366 @@
+
+    # run.runresults.finalize_save()
+
+    # out, _ = capsys.readouterr()
```
This allows Python to look at the stdout from the run, so it can be kept.

peterscherpelz commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,366 @@
+    # Cleanup and final output                                            #
+    #######################################################################
+
+    # run.runresults.finalize_save()
```
Doesn't exist yet so can discard.

peterscherpelz commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,366 @@
+
+    # out, _ = capsys.readouterr()
+
+    # filelist = [
```
Everything from here down can probably be eliminated for now in this test function.

peterscherpelz commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,366 @@
+    #         "diags/histograms/particle_histogram_uzbirth_0000000106.pdf",
+    #     ]
+
+    # print(out)
```
Well you could keep some asserts like this that checks a few lines in the output are in fact present. ANd if there are output files that should be created you can check for them.

peterscherpelz reviewed (commented):
More comments. Everything is still on the right track; they key is to fold as much as possible into the diode setup rather than the test function. Most of the previous analysis doesn't exist yet in WarpX so we can remove that.

Commit:
Calcualte RES_LENGTH if NX, NY, NZ are passed in

Commit:
Species and field diagnostics now defined in diode_setup

- Particle species can be passed in as an input argument
- The list of field diagnostic data to be tracked can be defined as
  an input parameter.

Commit:
uncommented and started porting directory functions

Commit:
Passed dimension and rz to diode template, removed unused stuff

peterscherpelz committed:
Significantly refactor testing and diode setup

KZhu-ME commented on mewarpx/mewarpx/util.py:
```diff
@@ -30,3 +31,18 @@ def init_libwarpx(ndim, rz):
+        os.makedirs(path)
+    except OSError as exc:
+        if exc.errno == errno.EEXIST and os.path.isdir(path):
+            pass
```
It looks like if the test directory with the same name already exists, you'll just keep using that directory. Would it be safer to truncate that directory in case the previously ran test had a file that won't be over written? Another alternative is to call a method that would re-generate and broadcast a new test directory name.

peterscherpelz replied:
I actually don't worry about collisions - the function is written this way primarily so that if `test_files` directory didn't exist it will be created rather than erroring out. Though I think `temp_files`  is a better name, so I just changed that... We've used it in other cases like making folders for diagnostics to go in; you don't have to worry if the parent directory of a diagnostic does or doesn't exist.

KZhu-ME commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,87 @@
+    # Initialize each run with consistent, randomly-chosen, rseed. Use a random
+    # seed instead for initial dataframe generation.
+    # np.random.seed()
+    np.random.seed(92160881)
```
To avoid collisions, you could use the current unix time as the seed, that would save having to manually set the seed each run.

peterscherpelz replied:
This seed is for random numbers drawn in the physics (admittedly in the python side only), not for the directories! The latter explicitly don't use this seed. So having this seed constant is nice as calculations can be exactly the same.

KZhu-ME commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,87 @@
+        V_ANODE_EXPRESSION="%.1f*sin(2*pi*%.5e*t)" % (VOLTAGE, FREQ),
+        D_CA=0.067,
+        INERT_GAS_TYPE='He',
+        N_INERT=9.64e20,  # m^-3
+        T_INERT=300.0,  # K
```
I believe that only P_INERT, and T_INERT will be passed from the PICMI script, and we will have to calculate N_INERT from those using the ideal gas law. I've implemented the ideal gas law function in `mewarpx/mewarpx/util.py` on the `kzhu/feature_mcc-wrapper` branch. The function is used in `mewarpx/mewarpx/mcc_wrapper.py`

peterscherpelz replied:
Yeah I left a comment on your PR suggesting we add N_INERT as an input option. So I'll throw this one back on you ;-)

KZhu-ME reviewed (commented):
Overall looks good!

Commit:
fixed several issues

peterscherpelz committed:
Fix issues in moving from diode along x to along z

Also force PERIOD to be specified at present, and make RES_LENGTH always
reflect z direction.

peterscherpelz committed:
Make tests run in `temp_files`, not `test_files`.

peterscherpelz commented:
@MKieburtz OK, I think I found and fixed the problem. We hadn't completely swapped things between x and z, such as NX and NZ which are now [16, 128] like they should be in the new setup.

I think the acute problem was specifying `warpx_potential_lo_x` rather than `warpx_potential_lo_z` - at that point things worked again.

From my point of view we're good to go, though I haven't actually verified output is correct, just that it doesn't crash.

EtherealMC-Bit commented:
Interesting! That's good news for sure

peterscherpelz committed:
Remove old test no longer needed.

EtherealMC-Bit commented:
Can I make this PR not a draft?

Commit:
Initialized the simulation inside mwxrun instead of diode_setup

peterscherpelz commented:
> Can I make this PR not a draft?

For me the button is right above the check status, just below all the comments. "Ready for review"

peterscherpelz committed:
Remove passing simulation to init_run.

Also fixed test_capacitive_discharge.py for full compatibility with
pytest, and check that a string from run completion is present.

roelof-groenewald commented on mewarpx/tests/test_capacitive_discharge.py:
```diff
@@ -0,0 +1,98 @@
+        NX=16,
+        NZ=128,
+        # This gives equal spacing in x & z
+        PERIOD=D_CA * 16 / 128.0,
```
```suggestion
        PERIOD=D_CA * NX / NY,
```

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -33,16 +33,16 @@ class MEWarpXRun(object):
 
     def __init__(self):
         self.initialized = False
+        self.simulation = picmi.Simulation()
```
```suggestion
        self.simulation = picmi.Simulation(verbose=0)
```

roelof-groenewald reviewed (approved):
@peterscherpelz or @MKieburtz I made two quick suggestions for changes I think would be good. I'll approve this anyway so you guys can decide whether to accept those suggestions or to merge without them.

Commit:
Suggested changes

- Initialized the picmi simulation with verbosity=0
- Defined variables for NX and NZ instead of reusing the numbers

EtherealMC-Bit commented:
I think we're good to go @roelof-groenewald

KZhu-ME reviewed (approved):
Everything looks good!
